openfe/protocols/restraint_utils/geometry/boresch.py

```diff
@@ -0,0 +1,870 @@
+        # representative of the input and go with it
+        guest_ag = u.select_atoms[guest_idxs]
+        guest_anchor = [
+            at.ix for at in guest_ag.atoms[guest_restraint_atoms_idxs]
+        ]
+        host_ag = u.select_atoms[host_idxs]
+        host_anchor = [
+            at.ix for at in host_ag.atoms[host_restraint_atoms_idxs]
+        ]
+
+        # Set the equilibrium values as those of the final frame
+        u.trajectory[-1]
+        atomgroup = u.atoms[host_anchor + guest_anchor]
+        bond, ang1, ang2, dih1, dih2, dih3 = _get_restraint_distances(
+            atomgroup
+        )
+
+        return BoreschRestraintGeometry(
+            host_atoms=host_anchor,
+            guest_atoms=guest_anchor,
+            r_aA0=bond,
+            theta_A0=ang1,
+            theta_B0=ang2,
+            phi_A0=dih1,
+            phi_B0=dih2,
+            phi_C0=dih3
+        )
+
+    if (guest_restraint_atoms_idxs is not None) ^ (host_restraint_atoms_idxs is not None):  # fmt: skip
+        # This is not an intended outcome, crash out here
+        errmsg = (
+            "both ``guest_restraints_atoms_idxs`` and "
+            "``host_restraint_atoms_idxs`` "
+            "must be set or both must be None. "
+            f"Got {guest_restraint_atoms_idxs} and {host_restraint_atoms_idxs}"
+        )
+        raise ValueError(errmsg)
+
+    # 1. Fetch the guest anchors
+    guest_anchors = find_guest_atom_candidates(
+        topology=topology,
+        trajectory=trajectory,
+        rdmol=guest_rdmol,
+        guest_idxs=guest_idxs,
+        rmsf_cutoff=rmsf_cutoff,
+    )
+
+    if len(guest_anchors) != 0:
+        errmsg = "No suitable ligand atoms found for the restraint."
+        raise ValueError(errmsg)
+
+    # 2. We then loop through the guest anchors to find suitable host atoms
+    for guest_anchor in guest_anchors:
+        # We next fetch the host atom pool
+        host_pool = find_host_atom_candidates(
+            topology=topology,
+            trajectory=trajectory,
+            host_idxs=host_idxs,
+            l1_idx=guest_anchor,
+            host_selection=host_selection,
+            dssp_filter=dssp_filter,
+            rmsf_cutoff=rmsf_cutoff,
+            min_distance=host_min_distance,
+            max_distance=host_max_distance,
+        )
+
+        host_anchor = _find_host_anchor(
+            guest_atoms=u.atoms[list(guest_anchor)],
+            host_atom_pool=u.atoms[host_pool],
+            minimum_distance=0.5 * unit.nanometer,
+            angle_force_constant=angle_force_constant,
+            temperature=temperature,
+        )
+        # continue if it's empty, otherwise stop
+        if host_anchor is not None:
+            break
+
+    if host_anchor is None:
+        errmsg = "No suitable host atoms could be found"
+        raise ValueError(errmsg)
+
+    # Set the equilibrium values as those of the final frame
+    u.trajectory[-1]
```

Comment:
Raise an issue / todo here to experiment on what frame / values to pick - ideally some kind of mean with the frame closest to the mean values.